### PR TITLE
Implement tolerance fill transform on OwidTable

### PR DESCRIPTION
**Pull request by danielgavrilov**
Working on applying them on charts on another branch.

### CODE REVIEW

File: coreTable/InvalidCells.ts [hunk 1]
```diff
@@ -36,6 +37,7 @@ export const InvalidCellTypes = {
     MissingValuePlaceholder: new MissingValuePlaceholder(),
     NotAParseableNumberButShouldBeNumber: new NotAParseableNumberButShouldBeNumber(),
     DivideByZeroError: new DivideByZeroError(),
+    NoValueWithinTolerance: new NoValueWithinTolerance(),
```
Comment (breck7) [inline coreTable/InvalidCells.ts:40]: +1

File: coreTable/LegacyToOwidTable.ts [hunk 2]
```diff
@@ -106,9 +107,7 @@ export const legacyVariablesToColDefsAndOwidRowsSortedByTimeAsc = (
             })
         }
 
-        const timeColumnName = columnDef.isDailyMeasurement
-            ? OwidTableSlugs.day
-            : OwidTableSlugs.year
+        const timeColumnName = timeColumnSlugFromColumnDef(columnDef)
```
Comment (breck7) [inline coreTable/LegacyToOwidTable.ts:110]: nit: I should have called "name" => "slug"

File: coreTable/OwidTable.test.ts [hunk 3]
```diff
@@ -451,3 +479,62 @@ describe("time domain", () => {
...
+1,2005,usa,1,`,
+            [
+                { slug: "gdp", type: ColumnTypeNames.Numeric },
+                { slug: "year", type: ColumnTypeNames.Year },
+            ]
+        )
+
+        const toleranceTable = table.fillColumnWithTolerance("gdp", 1)
+
+        expect(toleranceTable.rows).toEqual(
```
Comment (breck7) [inline coreTable/OwidTable.test.ts:501]: this is some next level jesting :)

File: coreTable/OwidTableUtil.ts [hunk 4]
```diff
@@ -0,0 +1,10 @@
+import { ColumnSlug } from "./CoreTableConstants"
+import { OwidColumnDef, OwidTableSlugs } from "./OwidTableConstants"
+
+export function timeColumnSlugFromColumnDef(def: OwidColumnDef) {
+    return def.isDailyMeasurement ? OwidTableSlugs.day : OwidTableSlugs.year
+}
+
+export function makeOriginalTimeSlugFromColumnSlug(slug: ColumnSlug) {
```
Comment (breck7) [inline coreTable/OwidTableUtil.ts:8]: nits:

Arrow functions? Reduce your code golf scores by 66% :)

```typescript
export const makeOriginalTimeSlugFromColumnSlug = (slug: ColumnSlug) => `${slug}-originalTime`
```

File: coreTable/OwidTable.ts [hunk 5]
```diff
@@ -577,6 +588,72 @@ export class OwidTable extends CoreTable<OwidRow, OwidColumnDef> {
...
+        const column = this.get(columnSlug)
+        const columnDef = column?.def as OwidColumnDef
+        const timeColumnSlug = timeColumnSlugFromColumnDef(columnDef)
+        const timeColumnDef = this.get(timeColumnSlug)?.def as OwidColumnDef
+        const originalTimeSlug = makeOriginalTimeSlugFromColumnSlug(columnSlug)
+
+        const originalRows = this.sortedByTime.rows
+        const allTimes = originalRows.map((row) => row[timeColumnSlug])
+
+        const rows = flatten(
```
Comment (breck7) [inline coreTable/OwidTable.ts:601]: One idea: What about splitting this method into 2 transforms? Something like `tableWhereEveryEntityGetsATimeForAllTimes()` and then the Interpolate from closest as a separate transform on that?

File: coreTable/OwidTable.ts [hunk 6]
```diff
@@ -266,24 +272,29 @@ export class OwidTable extends CoreTable<OwidRow, OwidColumnDef> {
         }, `Filter out countries with population less than ${minPop}`)
     }
 
-    filterByTargetTime(targetTime: Time, tolerance: Integer) {
+    filterByTargetTimes(targetTimes: Time[], tolerance: Integer = 0) {
```
Comment (breck7) [inline coreTable/OwidTable.ts:275]: I'm not seeing this used yet. Is this in anticipation of scatters?

File: coreTable/OwidTable.ts (hunk 5 above)
Comment (breck7) [reply coreTable/OwidTable.ts:601]: Should it be called `interpolateColumnWithTolerance`? My thinking is 1) we may add other (unrelated) interpolation functions in the future and might be helpful to find them. 2) In `grapher/stackedCharts/StackedUtils.ts` I actually think we could ditch those functions and then just fire up a CoreTable and reuse some of this code.

File: grapher/utils/Util.ts [hunk 7]
```diff
@@ -1063,7 +1070,84 @@ export function sortNumeric<T>(
...
 // A predicate for filtering an array of nulls and undefineds that returns the correct type
 export const isPresent = <T>(t: T | undefined | null | void): t is T =>
     t !== undefined && t !== null
+
+export function fillUndefinedWithClosest<
+    ValueSlug extends ColumnSlug,
+    TimeSlug extends ColumnSlug,
+    Row extends { [key in TimeSlug]?: number } & { [key in ValueSlug]?: any }
+>(
+    rows: Row[],
```
Comment (breck7) [inline grapher/utils/Util.ts:1087]: Looks good.

Does this assume rows are sorted by time asc, right? Maybe rename the `rows` argument to `rowsSortedByTimeAsc`?

Sometimes I've passed unsorted values into utils and gotten unexpected results.

**Review by breck7 — APPROVED** [review]
LGTM.
The `fillColumnWithTolerance` method seems like it could be two methods.

File: coreTable/OwidTable.ts (hunk 6 above)
Comment (danielgavrilov) [reply coreTable/OwidTable.ts:276]: Yeah, tables & slope charts, where you have two points – start & end.

Comment (danielgavrilov) [reply coreTable/OwidTable.ts:276]: Coming in `tolerance-charts` but left the utils in this one

File: coreTable/OwidTable.ts (hunk 5 above)
Comment (danielgavrilov) [reply coreTable/OwidTable.ts:619]: Yeah that's a good idea! And easier to debug. 

I extracted `injectAllTimeRowsForEveryEntity()` as a separate transform. Feel free to rename it!

File: coreTable/OwidTableUtil.ts (hunk 4 above)
Comment (danielgavrilov) [reply coreTable/OwidTableUtil.ts:8]: arrows are for arqueros

File: coreTable/OwidTable.test.ts (hunk 3 above)
Comment (danielgavrilov) [reply coreTable/OwidTable.test.ts:501]: Yeah I scrolled through some stackoverflow. :) 

We should maybe add some Jest utils, just mentioned it on Slack.

**Pull request merged**